File: articles/statistics-uncertainty-measurement-biology/r/descriptive_uncertainty.R
# Descriptive uncertainty workflow in R.

data_path <- file.path("data", "measurements.csv")
if (!file.exists(data_path)) {
  data_path <- file.path("..", "data", "measurements.csv")
}

data <- read.csv(data_path)

summarize_group <- function(x) {
  n <- length(x)
  mean_value <- mean(x)
  sd_value <- sd(x)
  se_value <- sd_value / sqrt(n)
  t_crit <- qt(0.975, df = n - 1)

  data.frame(
    n = n,
    mean = mean_value,
    standard_deviation = sd_value,
    standard_error = se_value,
    ci_lower = mean_value - t_crit * se_value,
    ci_upper = mean_value + t_crit * se_value
  )
}

rows <- lapply(split(data$value, data$group), summarize_group)

summary_df <- data.frame(
  group = names(rows),
  do.call(rbind, rows),
  row.names = NULL
)

print(round(summary_df, 5))

File: articles/statistics-uncertainty-measurement-biology/python/descriptive_uncertainty.py
"""
Descriptive uncertainty workflow.

Run:
    python python/descriptive_uncertainty.py
"""

from pathlib import Path

import pandas as pd

from statistics_measurement_core import descriptive_uncertainty


ARTICLE_DIR = Path(__file__).resolve().parents[1]
DATA_PATH = ARTICLE_DIR / "data" / "measurements.csv"


def main() -> None:
    data = pd.read_csv(DATA_PATH)

    rows = []

    for group, group_df in data.groupby("group"):
        summary = descriptive_uncertainty(group_df["value"])

        rows.append(
            {
                "group": group,
                "n": summary.n,
                "mean": summary.mean,
                "standard_deviation": summary.standard_deviation,
                "standard_error": summary.standard_error,
                "ci_lower": summary.ci_lower,
                "ci_upper": summary.ci_upper,
            }
        )

    print(pd.DataFrame(rows).round(5).to_string(index=False))


if __name__ == "__main__":
    main()
